shell/platform/common/cpp/client_wrapper/include/flutter/encodable_value.h

```diff
@@ -6,16 +6,150 @@
+//                                      EncodableValue(4),
+//                                  })},
+//   })
+//
+// The primary API surface for this object is std::variant. For instance,
+// getting a string value from an EncodableValue, with type checking:
+//   if (std::get<std::string>(value)) {
+//     std::string some_string = std::get<std::string>(value);
+//   }
+//
+// The order/indexes of the variant types is part of the API surface, and is
+// guaranteed not to change.
+class EncodableValue : public std::variant<std::monostate,
+                                           bool,
+                                           int32_t,
+                                           int64_t,
+                                           double,
+                                           std::string,
+                                           std::vector<uint8_t>,
+                                           std::vector<int32_t>,
+                                           std::vector<int64_t>,
+                                           std::vector<double>,
+                                           EncodableList,
+                                           EncodableMap> {
+ public:
+  // Rely on std::variant for most of the constructors/operators.
+  using super = std::variant<std::monostate,
+                             bool,
+                             int32_t,
+                             int64_t,
+                             double,
+                             std::string,
+                             std::vector<uint8_t>,
+                             std::vector<int32_t>,
+                             std::vector<int64_t>,
+                             std::vector<double>,
+                             EncodableList,
+                             EncodableMap>;
+  using super::super;
+  using super::operator=;
+
+  explicit EncodableValue() : super() {}
+
+  // Avoid the C++17 pitfall of conversion from char* to bool. Should not be
+  // needed for C++20.
+  explicit EncodableValue(const char* string) : super(std::string(string)) {}
+  EncodableValue& operator=(const char* other) {
+    *this = std::string(other);
+    return *this;
+  }
+
+  // Override the conversion constructors from std::variant to make them
+  // explicit, to avoid implicit conversion.
+  //
+  // While implicit conversion can be convenient in some cases, it can have very
+  // surprising effects. E.g., calling a function that takes an EncodableValue
+  // but accidentally passing an EncodableValue* would, instead of failing to
+  // compile, go through a pointer->bool->EncodableValue(bool) chain and
+  // silently call the function with a temp-constructed EncodableValue(true).
+  template <class T>
+  constexpr explicit EncodableValue(T&& t) noexcept : super(t) {}
+
+  // Returns true if the value is null. Convenience wrapper since unlike the
+  // other types, std::monostate uses aren't self-documenting.
+  bool IsNull() const { return index() == 0; }
```

Comment:
That the index of the monotstate is 0 is not immediately obvious (though I realize you have documented it as being stable and part of the API). How about just `std:holds_alternative<std::monostate>(...)` ?

Reply:
Oops, I meant to change this back. I started with that more obvious implementation, but tried this while I was trying to figure out why IsNull was giving me weird results in a unit test (which turned out eventually to be the pointer->bool->EncodableValue(bool) implicit conversion chain with an accidental call-with-pointer when this was freestanding.)